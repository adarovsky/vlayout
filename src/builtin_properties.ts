import { TypeDefinition } from './types';
import { EMPTY, Observable } from 'rxjs';
import { Expression } from './expression';

export class BuiltinProperty extends Expression {
    constructor(readonly keyPath: string, readonly typeDefinition: TypeDefinition, readonly sink: Observable<any>) {
        super(0, 0);
    }

    instantiate(): this {
        const v = new (this.constructor as typeof BuiltinProperty)(this.keyPath, this.typeDefinition, this.sink);
        return v as this;
    }

    toString(): string {
        return this.keyPath;
    }
}
