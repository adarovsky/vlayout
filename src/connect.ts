import React, { createElement } from 'react';
import { combineLatest, Observable, of, Subscription } from 'rxjs';
import { fromPairs, toPairs, unzip, zip } from 'lodash';

export type ObservableMapObject<S = any> = {
    [K in keyof S]: Observable<S[K]>|S[K]
}

export function connect<P, S = {}>(WrappedComponent: React.FC<P> | React.ComponentClass<P, S>) {
    return class Connector extends React.PureComponent<ObservableMapObject<P>, P> {
        subscription = new Subscription();

        componentDidMount(): void {
            this.subscription.unsubscribe();
            const pairs = toPairs(this.props);
            const [keys, values] = unzip(pairs);
            this.subscription = combineLatest(values.map(v => v instanceof Observable ? v : of(v))).subscribe(s => {
                const pairs = zip(keys, s);
                const state = fromPairs(pairs);
                this.setState(state as P);
            });
        }

        componentWillUnmount(): void {
            this.subscription.unsubscribe();
        }

        render(): React.ReactElement<any, string | React.JSXElementConstructor<any>> | string | number | {} | React.ReactNodeArray | React.ReactPortal | boolean | null | undefined {
            if (this.state) {
                return createElement(WrappedComponent, this.state);
            } else {
                return null;
            }
        }
    }
}
