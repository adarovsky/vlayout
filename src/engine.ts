import { Inputs } from './inputs';
import {
    Dictionary,
    EnumDefinition,
    ListDefinition,
    ListDefinitionItem,
    TypeDefinition,
    Types,
} from './types';
import { EnumValue, Expression } from './expression';
import {
    Ceil,
    ColorAlpha,
    Floor,
    FontFamily,
    FontSized,
    FontTyped,
    FunctionImplementationI,
    Image,
    LocalizedNumber,
    LocalizedString,
    Max,
    Min,
    Round,
    ShortLocalizedNumber,
    StringPrefix,
    StringSuffix,
    SubString,
    ToCapitalCase,
    ToLower,
    ToUpper,
} from './builtin_functions';
import React from 'react';
import { ViewListReference, ViewReference } from './view_reference';
import { View } from './view';
import { ReactViewProps } from './react_views';
import { Button, TextField } from './primitives';
import { Observable } from 'rxjs';
import {
    ListClickHandler,
    ListEnterHandler,
    ListModelItem,
    ListTextChangeHandler,
} from './list';
import { isEqual } from 'lodash';
import invariant from 'ts-invariant';

export class Engine {
    readonly inputs = new Inputs(this);
    readonly types = new Types(this);
    readonly functions: FunctionImplementationI[];
    private readonly referencedViews: Dictionary<ViewReference> = {};
    private readonly buttons: Dictionary<() => Promise<void>> = {};
    private readonly textFields: Dictionary<{
        onChange: (s: string) => void;
        onEnter: () => void;
    }> = {};
    private readonly listButtons: Dictionary<ListClickHandler> = {};
    private readonly listTextFields: Dictionary<{
        onChange: ListTextChangeHandler;
        onEnter: ListEnterHandler;
    }> = {};
    private readonly referencedListViews: Dictionary<ViewListReference> = {};
    readonly valueSnapshot: {
        inputs: Dictionary<any>;
        properties: Dictionary<any>;
    } = {
        inputs: {},
        properties: {},
    };

    constructor(readonly debug = false, readonly verboseIds: string[] = []) {
        this.functions = [
            new LocalizedNumber(this),
            new ShortLocalizedNumber(this),
            new Image(this),
            new FontFamily(this),
            new FontTyped(this),
            new FontSized(this),
            new ColorAlpha(this),
            new StringPrefix(this),
            new StringSuffix(this),
            new SubString(this),
            new ToUpper(this),
            new ToLower(this),
            new ToCapitalCase(this),
            new Round(this),
            new Floor(this),
            new Ceil(this),
        ];
    }

    type(name: string): TypeDefinition | null {
        return this.types.type(name);
    }

    boolType(): TypeDefinition {
        return this.type('Bool')!;
    }

    numberType(): TypeDefinition {
        return this.type('Number')!;
    }

    colorType(): TypeDefinition {
        return this.type('Color')!;
    }

    stringType(): TypeDefinition {
        return this.type('String')!;
    }

    fontType(): TypeDefinition {
        return this.type('Font')!;
    }

    imageType(): TypeDefinition {
        return this.type('Image')!;
    }

    variableForKeyPath(keyPath: string): Expression | null {
        const kp = keyPath.split('.');
        if (kp.length === 2) {
            const enumType = this.type(kp[0]);
            if (enumType instanceof EnumDefinition) {
                const v = new EnumValue(kp[1], 0, 0);
                v.linkEnum(enumType);
                return v;
            }
        }

        return this.inputs.input(keyPath, 0, 0);
    }

    functionFor(
        name: string,
        parameters: TypeDefinition[]
    ): FunctionImplementationI {
        if (name === '@') {
            for (let p of parameters) {
                if (p !== this.stringType()) {
                    throw new Error(`localized string receives only strings`);
                }
            }

            return new LocalizedString(this);
        } else if (name === 'Min') {
            for (let p of parameters) {
                if (p !== this.numberType()) {
                    throw new Error(`Min accepts only numbers`);
                }
            }
            if (parameters.length == 0) {
                throw new Error(`Not enough arguments to Min`);
            }

            return new Min(this);
        } else if (name === 'Max') {
            for (let p of parameters) {
                if (p !== this.numberType()) {
                    throw new Error(`Max accepts only numbers`);
                }
            }

            if (parameters.length == 0) {
                throw new Error(`Not enough arguments to Max`);
            }

            return new Max(this);
        } else {
            for (let f of this.functions) {
                if (f.name === name && isEqual(f.parameterTypes, parameters)) {
                    return f;
                }
            }

            throw new Error(
                `function ${name}(${parameters
                    .map((p) => p.toString())
                    .join(', ')}) not found`
            );
        }
    }

    functionsLoose(
        name: string,
        parametersCount: number
    ): FunctionImplementationI[] {
        return this.functions.filter(
            (f) =>
                f.name === name && f.parameterTypes.length === parametersCount
        );
    }

    registerView(
        key: string,
        createComponent: (parent: View) => React.ReactElement<ReactViewProps>
    ) {
        invariant(
            typeof createComponent === 'function',
            `createComponent for ${key} is empty`
        );
        this.referencedViews[key] = new ViewReference(createComponent);
    }

    registerInput(
        name: string,
        type: TypeDefinition,
        sink: Observable<any>
    ): void {
        invariant(sink, `input sink for ${name} is empty`);
        this.inputs.registerInput(name, type, sink);
    }

    registerButton(key: string, onClick: () => Promise<void>) {
        this.buttons[key] = onClick;
    }

    registerTextField(
        key: string,
        onChange: (s: string) => void,
        sink: Observable<string>,
        onEnter: () => void = () => {}
    ) {
        this.textFields[key] = { onChange, onEnter };
        this.inputs.registerInput(key, this.stringType(), sink);
    }

    registerEnum(name: string, values: Dictionary<any>): void {
        this.types.registerEnum(new EnumDefinition(this, name, values));
    }

    registerStandardEnum<EnumType>(
        name: string,
        e: StandardEnum<EnumType>
    ): void {
        this.types.registerEnum(
            new EnumDefinition(this, name, materializeEnum(e))
        );
    }

    registerList(name: string, fields: Dictionary<ListDefinitionItem>): void {
        this.types.registerList(new ListDefinition(this, name, fields));
    }

    registerListButton(
        name: string,
        onClick: (item: ListModelItem) => Promise<void>
    ) {
        this.listButtons[name] = onClick;
    }

    registerListTextField(
        name: string,
        onChange: ListTextChangeHandler,
        onEnter: ListEnterHandler = () => {}
    ) {
        this.listTextFields[name] = { onChange, onEnter };
    }

    registerListView(
        name: string,
        createComponent: (
            parent: View,
            modelItem: Observable<ListModelItem>
        ) => React.ReactElement<ReactViewProps>
    ) {
        this.referencedListViews[name] = new ViewListReference(createComponent);
    }

    viewForKey(key: string): View | null {
        if (this.buttons[key]) return new Button(this.buttons[key]);
        else if (this.textFields[key])
            return new TextField(
                key,
                this.textFields[key].onChange,
                this.textFields[key].onEnter
            );
        else if (this.referencedViews[key])
            return new ViewReference(this.referencedViews[key].createComponent);
        else return null;
    }

    listButtonForKey(key: string): ListClickHandler | null {
        return this.listButtons[key] || null;
    }

    listTextFieldForKey(
        key: string
    ): { onChange: ListTextChangeHandler; onEnter: ListEnterHandler } | null {
        return this.listTextFields[key] || null;
    }

    listViewForKey(key: string): ViewListReference | null {
        const view = this.referencedListViews[key];
        if (view) {
            return new ViewListReference(view.createComponent);
        }

        return null;
    }

    logInputValue(name: string, value: any): void {
        this.valueSnapshot.inputs[name] = value;
    }

    beginUpdates() {
        this.inputs.beginUpdates();
    }

    endUpdates() {
        this.inputs.endUpdates();
    }
}

export type StandardEnum<T> = {
    [id: string]: T | string;
    [nu: number]: string;
};

function camelize(str: string) {
    return str
        .replace(/(?:^\w|[A-Z]|\b\w|-|\s+)/g, function (match, index) {
            if (+match === 0) return ''; // or if (/\s+/.test(match)) for white spaces
            return index === 0 ? match.toLowerCase() : match.toUpperCase();
        })
        .replace('-', '');
}

function materializeEnum<EnumType>(e: StandardEnum<EnumType>) {
    return Object.keys(e)
        .filter((x) => isNaN(+x))
        .reduce((prev, cur) => ({ ...prev, [camelize(cur)]: e[cur] }), {});
}
