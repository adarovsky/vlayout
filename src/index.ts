export {Engine} from "./engine";
export {Layout} from "./layout";
export {ReactView, ReactViewState} from "./react_views";
import './vlayout.css';
