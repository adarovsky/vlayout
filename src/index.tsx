export {
    resizeObserver,
    pauseObserving,
    resumeObserving,
} from './resize_sensor';
export { Engine } from './engine';
export { Layout } from './layout';
import './vlayout.css';

export { ReactViewReference as ReactView } from './react_views';
export type { ReactViewProps, ReactViewState } from './react_views';
export { View } from './view';
export { ColorContainer as Color, ImageContainer as Image } from './types';
export type {
    SimpleListDefinitionItem,
    ListDefinitionItem,
    ComplexDefinitionItem,
} from './types';
export type { ListModelItem } from './list';
export { LayoutComponent } from './layout_component';
export { connect } from './connect';
export type { ObservableMapObject } from './connect';
