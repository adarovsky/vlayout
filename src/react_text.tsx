import {ViewProperty} from "./view";
import {combineLatest, of} from "rxjs";
import React, {CSSProperties} from "react";
import _ from "lodash";
import {fontStyle, ReactRoundRect} from "./react_primitives";
import {ColorContainer, Dictionary} from "./types";
import {TextField} from "./primitives";
import {ReactViewProps, ReactViewState} from "./react_views";

export interface ReactTextFieldState extends ReactViewState {
    text: string;
    placeholder: string;
    enabled: boolean;
    fontStyle: CSSProperties;
    colorStyle: CSSProperties;
    height: number;
    type: string;
}

export class ReactTextFieldBase<S extends ReactTextFieldState = ReactTextFieldState> extends ReactRoundRect<S> {
    readonly inputRef = React.createRef<HTMLInputElement>();

    constructor(props: ReactViewProps) {
        super(props);
        this.state = {...this.state,
            text: '',
            placeholder: '',
            enabled: true,
            fontStyle: {},
            colorStyle: {},
            height: 0,
            type: 'regular'
        };
    }


    public styleProperties(): ViewProperty[] {
        const sup = super.styleProperties();
        return sup.concat(this.props.parentView.activePropertiesNamed(
            'contentPadding.left', 'contentPadding.top', 'contentPadding.right', 'contentPadding.bottom'));
    }

    componentDidMount(): void {
        super.componentDidMount();

        this.wire('text', 'text', x => x);
        this.wire('enabled', 'enabled', x => x);
        this.wire('font', 'fontStyle', fontStyle);
        this.wire('textColor', 'colorStyle', x => ({color: x.toString()}));
        this.wire('type', 'type', x => x);
        const props = [this.intrinsicSize(),
            this.props.parentView.property('contentPadding.top').value?.sink ?? of(0),
            this.props.parentView.property('contentPadding.bottom').value?.sink ?? of(0),
            this.props.parentView.property('strokeWidth').value?.sink ?? of(0),
            this.props.parentView.property('strokeColor').value?.sink ?? of(null)];
        this.subscription.add(combineLatest(props).subscribe(x => {
            if (super.isHeightDefined()) {
                const size = x[0];
                const top = x[1] || 0 as number;
                const bottom = x[2] || 0 as number;
                const strokeWidth = x[3] || 0 as number;
                const strokeColor = x[4] as ColorContainer|null;
                const strokeOffset = (strokeWidth > 0 && strokeColor !== null) ? strokeWidth : 0;
                this.setState(s => ({...s, height: size.height - top - bottom - strokeOffset*2}));
            }
        }));
    }

    styleValue(props: ViewProperty[], value: any[]): React.CSSProperties {
        const r = super.styleValue(props, value);

        value.forEach((val, index) => {
            switch (props[index].name) {
                case 'contentPadding.left':
                    r.paddingLeft = `${val}px`;
                    break;
                case 'contentPadding.right':
                    r.paddingRight = `${val}px`;
                    break;
                case 'contentPadding.top':
                    r.paddingTop = `${val}px`;
                    break;
                case 'contentPadding.bottom':
                    r.paddingBottom = `${val}px`;
                    break;
                case 'alpha':
                    r.display = val > 0 ? undefined : 'none';
                    break;
            }
        });

        r.pointerEvents = 'auto';

        return r;
    }

    protected textEntered(e: string): void {
    }

    protected enterPressed(): void {
    }

    protected isHeightDefined(): boolean {
        return true;
    }

    private textFieldStyle(): CSSProperties {
        const r: CSSProperties = {minWidth: 0, width: '100%'};

        _.extend(r, this.state.fontStyle);
        _.extend(r, this.state.colorStyle);
        _.extend(r, _.pick(this.state.style,
            'borderColor', 'borderStyle', 'borderRadius', 'paddingLeft', 'paddingRight', 'paddingTop', 'paddingBottom'));
        if (!this.state.style.borderColor && !this.state.style.borderStyle && !this.state.style.borderRadius) {
            r.border = 'none';
        }
        r.boxSizing = 'border-box';
        r.padding = '0 0 0 0';
        if (this.state.height > 0) {
            r.lineHeight = this.state.height + 'px';
        }
        return r;
    }


    style(): React.CSSProperties {
        return _.omit(this.state.style,
            'borderColor', 'borderStyle', 'borderRadius', 'paddingLeft',
            'paddingRight', 'paddingTop', 'paddingBottom');
    }

    render() {
        let inputMode: Dictionary<string> = {};
        switch (this.state.type) {
            case 'regular':
                break;
            case 'go':
                inputMode.type='text';
                inputMode.title = 'go';
                break;
            case 'numeric':
                inputMode.type='text';
                inputMode.inputMode = 'numeric';
                break;
            case 'search':
                inputMode.type='search';
                inputMode.inputMode = 'search';
                break;
            case 'phone':
                inputMode.type='text';
                inputMode.inputMode = 'tel';
                break;

            case 'url':
                inputMode.type='text';
                inputMode.inputMode = 'url';
                break;
        }

        let input = <input style={this.textFieldStyle()}
                           {...inputMode as Dictionary<string>}
                           placeholder={this.state.placeholder}
                           value={this.state.text}
                           onChange={e => this.textEntered(e.target.value)}
                           onKeyPress={e => {if (e.key === 'Enter') this.enterPressed();}}
                           ref={this.inputRef}/>;
        if (this.state.type === 'go') {
            input = <form action={'#'}>{input}</form>;
        }
        const extra = _.pick(this.state, 'id');
        return (<div {...extra}
                     style={this.style()}
                     className={this.className}
                     ref={this.viewRef}>
            {input}
        </div>);
    }
}

export class ReactTextField extends ReactTextFieldBase {
    protected textEntered(e: string): void {
        (this.props.parentView as TextField).onChange(e);
    }

    protected enterPressed(): void {
        (this.props.parentView as TextField).onEnter();
    }
}
