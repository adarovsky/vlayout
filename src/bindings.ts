import { Engine } from 'engine';
import { View } from 'view';
import { Button, TextField } from 'primitives';

export class Bindings {
    constructor(public readonly engine: Engine) {

    }

    registerSlot(name: string, type: string): void {
        if (type === 'listButton' && this.engine.listButtonForKey(name))
            return;

        if (type === 'listTextField' && this.engine.listTextFieldForKey(name))
            return;

        if (type === 'listView' && this.engine.listViewForKey(name))
            return;

        const v = this.viewForKey(name);
        if (!v) {
            throw new Error(`view ${name} is not registered`);
        }

        if (type === 'button' && !(v instanceof Button)) {
            throw new Error(`view ${name} is not a button, but ${v.viewType()}`);
        }

        if (type === 'textField' && !(v instanceof TextField)) {
            throw new Error(`view ${name} is not a text field, but ${v.viewType()}`);
        }
    }

    viewForKey(key: string): View | null {
        return this.engine.viewForKey(key);
    }

    toString(): string {
        return 'bindings {\n}\n';
    }
}
