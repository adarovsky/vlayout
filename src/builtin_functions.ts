import {Engine} from "./engine";
import {ColorContainer, FontContainer, ImageContainer, TypeDefinition} from "./types";
import {combineLatest, Observable} from "rxjs";
import {map} from "rxjs/operators";

export interface FunctionImplementationI {
    returnType: TypeDefinition;
    name: string;
    parameterTypes: TypeDefinition[];

    sink(parameters: Observable<any>[]): Observable<any>;
}

export class FunctionImplementation {
    constructor(public readonly engine: Engine) {
    }
}

export class LocalizedNumber extends FunctionImplementation implements FunctionImplementationI {
    name: string;
    parameterTypes: TypeDefinition[];
    returnType: TypeDefinition;

    constructor(engine: Engine) {
        super(engine);

        this.name = 'String';
        this.parameterTypes = [engine.numberType()];
        this.returnType = engine.stringType();
    }

    sink(parameters: Observable<any>[]): Observable<any> {
        return parameters[0].pipe(
            map( x => `${x}`)
        );
    }
}

export class ShortLocalizedNumber extends FunctionImplementation implements FunctionImplementationI {
    name: string;
    parameterTypes: TypeDefinition[];
    returnType: TypeDefinition;

    constructor(engine: Engine) {
        super(engine);

        this.name = 'ShortString';
        this.parameterTypes = [engine.numberType()];
        this.returnType = engine.stringType();
    }

    sink(parameters: Observable<any>[]): Observable<any> {
        return parameters[0].pipe(
            map( x => {
                if (x > 1000000) {
                    return `${Math.round(x / 100000) / 10} M`
                }
                else if (x > 1000) {
                    return `${Math.round(x / 100) / 10} K`
                }
                return `${x}`;
            })
        );
    }
}

export class LocalizedString extends FunctionImplementation implements FunctionImplementationI {
    name: string;
    parameterTypes: TypeDefinition[];
    returnType: TypeDefinition;

    constructor(engine: Engine) {
        super(engine);

        this.name = 'String';
        this.parameterTypes = [engine.stringType()];
        this.returnType = engine.stringType();
    }

    sink(parameters: Observable<any>[]): Observable<any> {
        return combineLatest(parameters as Observable<string>[]).pipe(
            map(params => {
                let key = params[0];
                for (let i = params.length - 1; i >= 1; --i) {
                    key = key.replace(`$${i}`, params[i]);
                }

                return key;
            })
        );
    }
}

export class Image extends FunctionImplementation implements FunctionImplementationI {
    name: string;
    parameterTypes: TypeDefinition[];
    returnType: TypeDefinition;

    constructor(engine: Engine) {
        super(engine);

        this.name = 'Image';
        this.parameterTypes = [engine.stringType()];
        this.returnType = engine.imageType();
    }

    sink(parameters: Observable<any>[]): Observable<any> {
        return parameters[0].pipe(
            map(x => new ImageContainer(x))
        );
    }
}

export class FontFamily extends FunctionImplementation implements FunctionImplementationI {
    name: string;
    parameterTypes: TypeDefinition[];
    returnType: TypeDefinition;

    constructor(engine: Engine) {
        super(engine);

        this.name = 'Font';
        this.parameterTypes = [engine.stringType(), engine.numberType()];
        this.returnType = engine.fontType();
    }

    sink(parameters: Observable<any>[]): Observable<any> {
        return combineLatest(parameters as Observable<any>[]).pipe(
            map(params => new FontContainer(params[0], params[1], null))
        );
    }
}

export class FontTyped extends FunctionImplementation implements FunctionImplementationI {
    name: string;
    parameterTypes: TypeDefinition[];
    returnType: TypeDefinition;

    constructor(engine: Engine) {
        super(engine);

        this.name = 'Font';
        this.parameterTypes = [engine.type('FontType')!, engine.numberType()];
        this.returnType = engine.fontType();
    }

    sink(parameters: Observable<any>[]): Observable<any> {
        return combineLatest(parameters as Observable<any>[]).pipe(
            map(params => new FontContainer(null, params[1], params[0]))
        );
    }
}

export class FontSized extends FunctionImplementation implements FunctionImplementationI {
    name: string;
    parameterTypes: TypeDefinition[];
    returnType: TypeDefinition;

    constructor(engine: Engine) {
        super(engine);

        this.name = 'Font';
        this.parameterTypes = [engine.numberType()];
        this.returnType = engine.fontType();
    }

    sink(parameters: Observable<any>[]): Observable<any> {
        return parameters[0].pipe(
            map(params => new FontContainer(null, params, null))
        );
    }
}

export class ColorAlpha extends FunctionImplementation implements FunctionImplementationI {
    name: string;
    parameterTypes: TypeDefinition[];
    returnType: TypeDefinition;

    constructor(engine: Engine) {
        super(engine);

        this.name = 'Color';
        this.parameterTypes = [engine.colorType(), engine.numberType()];
        this.returnType = engine.colorType();
    }

    sink(parameters: Observable<any>[]): Observable<any> {
        return combineLatest(parameters as Observable<any>[]).pipe(
            map(params => {
                const original = params[0] as ColorContainer;
                const alpha = params[1] as number;
                return new ColorContainer(original.red, original.green, original.blue, Math.max(0, Math.min(alpha, 1)));
            })
        );
    }
}