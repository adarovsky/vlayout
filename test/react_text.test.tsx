import React from "react";
import {Engine, Layout} from "../src";
import {asyncScheduler, BehaviorSubject, of, scheduled, Subject} from "rxjs";
import {mount} from "enzyme";
import sinon from "sinon";
import {ReactTextField} from "../src/react_text";

let engine: Engine | null = null;
function wait() {
  return scheduled([], asyncScheduler).toPromise();
}

describe("text", () => {
  beforeEach(() => {
    engine = new Engine();
  });

  it("cleans up on unmount", async () => {
    const subject = new Subject<number>();
    engine!.registerInput("test", engine!.numberType(), subject);
    engine!.registerTextField("myText", async () => {}, of(''));
    const spy = sinon.spy(ReactTextField.prototype, "setState");
    subject.next(2);
    const wrapper = mount(
      <Layout
        engine={engine!}
        content={`
     bindings {
        myText: textField
     }
     inputs {test: Number}
     layout {
         layer {
             myText {
                 center { x : 0.5 y : 0.5 }
                 text : "test"
                 cornerRadius: test > 1 ? 0.5 : 0.1
                 backgroundColor: #cccccc
             }
         }
     }`}
      />
    );
    subject.next(3);

    const callCount = spy.callCount;

    wrapper.unmount();

    subject.next(0);

    await wait();

    expect(spy.callCount).toBe(callCount);
    spy.restore();
  });

  it("should contain id if set", async function() {
    const text = new BehaviorSubject('');
    engine!.registerTextField("myText", async () => {}, text);

    const wrapper = mount(
        <Layout
            engine={engine!}
            content={`
     layout {
         layer {
             myText {
                 id: "text1"
                 center { x: 0.5 y: 0.5 }
             }
         }
     }`}
        />
    );

    const node = wrapper.find(".vlayout_textField");

    expect(node.getDOMNode()).toMatchSnapshot();
  });

  it("should change value", async function() {
    const text = new BehaviorSubject('');
    engine!.registerTextField("myText", async () => {}, text);

    const wrapper = mount(
        <Layout
            engine={engine!}
            content={`
     layout {
         layer {
             myText {
                 id: "text1"
                 center { x: 0.5 y: 0.5 }
             }
         }
     }`}
        />
    );

    const node = wrapper.find(".vlayout_textField");

    expect(node.getDOMNode()).toMatchSnapshot();
    text.next('new value');

    expect(node.getDOMNode()).toMatchSnapshot();
  });

  it("should send changes back", async function() {
    const text = new BehaviorSubject('');
    engine!.registerTextField("myText", async (s: string) => text.next(s), text);

    const wrapper = mount(
        <Layout
            engine={engine!}
            content={`
     layout {
         layer {
             myText {
                 id: "text1"
                 center { x: 0.5 y: 0.5 }
             }
         }
     }`}
        />
    );

    const node = wrapper.find(".vlayout_textField");
    const input = node.find('input');
    input.simulate('change', {target: {value: 'entered text'}});

    expect(text.value).toBe('entered text');
  });

});
