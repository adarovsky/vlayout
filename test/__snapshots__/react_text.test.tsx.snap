// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`text should change value 1`] = `
<div
  class="vlayout_textField"
  id="text1"
  style="position: absolute; left: 50%; transform: translateX(-50%) translateY(-50%); z-index: 1; top: 50%; box-sizing: border-box; pointer-events: auto;"
>
  <input
    placeholder=""
    style="min-width: 0; width: 100%; box-sizing: border-box; padding: 0px 0px 0px 0px;"
    value=""
  />
  <div
    style="position: absolute; left: 0px; top: 0px; right: 0px; bottom: 0px; overflow: hidden; z-index: -1; visibility: hidden;"
  >
    <div
      style="position: absolute; left: 0px; top: 0px; width: 10000000px; height: 10000000px;"
    />
  </div>
  <div
    style="position: absolute; left: 0px; top: 0px; right: 0px; bottom: 0px; overflow: hidden; z-index: -1; visibility: hidden;"
  >
    <div
      style="position: absolute; left: 0px; top: 0px; width: 200%; height: 200%;"
    />
  </div>
</div>
`;

exports[`text should change value 2`] = `
<div
  class="vlayout_textField"
  id="text1"
  style="position: absolute; left: 50%; transform: translateX(-50%) translateY(-50%); z-index: 1; top: 50%; box-sizing: border-box; pointer-events: auto;"
>
  <input
    placeholder=""
    style="min-width: 0; width: 100%; box-sizing: border-box; padding: 0px 0px 0px 0px;"
    value="new value"
  />
  <div
    style="position: absolute; left: 0px; top: 0px; right: 0px; bottom: 0px; overflow: hidden; z-index: -1; visibility: hidden;"
  >
    <div
      style="position: absolute; left: 0px; top: 0px; width: 10000000px; height: 10000000px;"
    />
  </div>
  <div
    style="position: absolute; left: 0px; top: 0px; right: 0px; bottom: 0px; overflow: hidden; z-index: -1; visibility: hidden;"
  >
    <div
      style="position: absolute; left: 0px; top: 0px; width: 200%; height: 200%;"
    />
  </div>
</div>
`;

exports[`text should contain id if set 1`] = `
<div
  class="vlayout_textField"
  id="text1"
  style="position: absolute; left: 50%; transform: translateX(-50%) translateY(-50%); z-index: 1; top: 50%; box-sizing: border-box; pointer-events: auto;"
>
  <input
    placeholder=""
    style="min-width: 0; width: 100%; box-sizing: border-box; padding: 0px 0px 0px 0px;"
    value=""
  />
  <div
    style="position: absolute; left: 0px; top: 0px; right: 0px; bottom: 0px; overflow: hidden; z-index: -1; visibility: hidden;"
  >
    <div
      style="position: absolute; left: 0px; top: 0px; width: 10000000px; height: 10000000px;"
    />
  </div>
  <div
    style="position: absolute; left: 0px; top: 0px; right: 0px; bottom: 0px; overflow: hidden; z-index: -1; visibility: hidden;"
  >
    <div
      style="position: absolute; left: 0px; top: 0px; width: 200%; height: 200%;"
    />
  </div>
</div>
`;
